docs(allocator): correct and expand safety comments on `Allocator` methods

overlookmotel opened the pull request:
Correct safety comments on unsafe methods of `Allocator`, as suggested by AI in https://github.com/oxc-project/oxc/pull/13134#pullrequestreview-3125823899.

Also expand the safety comments for `Allocator::set_data_ptr`, and add comments on how its invariants are upheld at call sites.

Also correct typos and improve comment formatting.

overlookmotel commented:
* **#13139** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/13139?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/13139?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=13139).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented on crates/oxc_allocator/src/from_raw_parts.rs:
```diff
@@ -142,8 +142,7 @@ impl Allocator {
     /// 1. Set the data pointer back to its correct value before it is dropped, using [`set_data_ptr`].
-    /// 2. Wrap the `Allocator` in `ManuallyDrop`, and taking care of deallocating it manually
-    ///    with the correct pointer.
+    /// 2. Wrap the `Allocator` in `ManuallyDrop`, and its deallocate memory manually with the correct pointer.
```
Fixed.

overlookmotel commented on crates/oxc_allocator/src/from_raw_parts.rs:
```diff
@@ -204,15 +205,28 @@ impl Allocator {
+    /// because data pointer no longer correctly describes the start of the allocation obtained
+    /// from system allocator. If the `Allocator` were dropped, it'd be UB.
+    ///
+    /// Only use this method if you prevent that possibililty. e.g.:
```
Fixed.

Dunqing commented:
### Merge activity

* **Aug 17, 1:23 PM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
docs(allocator): correct and expand safety comments on `Allocator` methods (#13139)

Correct safety comments on unsafe methods of `Allocator`, as suggested by AI in https://github.com/oxc-project/oxc/pull/13134#pullrequestreview-3125823899.

Also expand the safety comments for `Allocator::set_data_ptr`, and add comments on how its invariants are upheld at call sites.

Also correct typos and improve comment formatting.
